File: api-code-examples/python/doc-delete.py
import iroh
import asyncio


async def main():
    node = await iroh.IrohNode.memory()

    # Real programs handle errors!
    author = await node.author_create()
    print(f"Created author {author}")

    doc = await node.doc_create()
    print(f"Created document {doc.id()}")

    key = b"python"
    hash = await doc.set_bytes(author, key, b"says hello")
    print(f"Inserted {hash}")

    # Get all the entries with default filtering and sorting
    query = iroh.Query.all(None)
    entries = await doc.get_many(query)

    print("Keys:")
    for entry in entries:
        key = entry.key()
        hash = entry.content_hash()
        content = await entry.content_bytes(doc)
        print(f'{key.decode("utf-8")} : {content.decode("utf-8")} (hash: {hash})')

        print(f"Removing entry for author {author} and prefix {key.decode('utf-8')}.")

    # Removes all entries from that author and with the prefix "key"
    # TODO: Currently Broken
    # num_removed = doc.del(author, key)
    # print(f"Removed {num_removed} entry")

    entries = await doc.get_many(query)

    print("Keys:")
    for entry in entries:
        key = entry.key()
        hash = entry.content_hash()
        content = await entry.content_bytes(doc)
        print(f'{key.decode("utf-8")} : {content.decode("utf-8")} (hash: {hash})')

# Output:
# Created author ybkptbq4imifxaj544hl5etyszhecuepp66qlezov7sdzm3hqk4a
# Created document ipqqeughovjrvcxl5sji3hlwycheqqgiajq5hgnf6vtqp6qigm6q
# Inserted bafkr4ihasgdyqs6onufsjrmk5h5vcg2ud75u2iaokavwiulyg7wfno6fte
# Keys:
# python : says hello (hash: bafkr4ihasgdyqs6onufsjrmk5h5vcg2ud75u2iaokavwiulyg7wfno6fte)
# Removing entry for author ybkptbq4imifxaj544hl5etyszhecuepp66qlezov7sdzm3hqk4a and prefix python.
# Removed 1 entry
# Keys:

asyncio.run(main())

File: api-code-examples/go/doc-delete.go
package main

import (
	"fmt"

	"github.com/n0-computer/iroh-ffi/iroh-go/iroh"
)

func main() {
	node, err := iroh.NewIrohNode("iroh_data_dir")
	if err != nil {
		// real programs handle errors!
		panic(err)
	}

	author, err := node.AuthorCreate()
	if err != nil {
		panic(err)
	}
	fmt.Printf("Created author %s\n", author.ToString())

	doc, err := node.DocCreate()
	if err != nil {
		panic(err)
	}
	fmt.Printf("Created document %s\n", doc.Id().ToString())

	key := []byte("go")
	hash, err := doc.SetBytes(author, key, []byte("says hello"))
	if err != nil {
		panic(err)
	}
	fmt.Printf("Inserted %s\n", hash.ToString())

	// get all the entries with default filtering and sorting
	query := iroh.QueryAll(nil)
	entries, err := doc.GetMany(query)
	if err != nil {
		panic(err)
	}

	fmt.Println("Keys:")
	for _, entry := range entries {
		key := entry.Key()
		hash := entry.ContentHash()
		content, err := doc.ReadToBytes(entry)
		if err != nil {
			panic(err)
		}
		fmt.Printf("%q : %q (hash: %s)\n", string(key), string(content), hash.ToString())
	}

	fmt.Printf("Removing entry for author %s and prefix %s.\n", author.ToString(), string(key))
	// removes all entries from that author and with the prefix "key"
	num_removed, err := doc.Delete(author, key)
	if err != nil {
		panic(err)
	}
	fmt.Printf("Removed %d entry\n", num_removed)

	// get all the entries with default filtering and sorting
	entries, err = doc.GetMany(query)
	if err != nil {
		panic(err)
	}

	fmt.Println("Keys:")
	for _, entry := range entries {
		key := entry.Key()
		hash := entry.ContentHash()
		content, err := doc.ReadToBytes(entry)
		if err != nil {
			panic(err)
		}
		fmt.Printf("%q : %q (hash: %s)\n", string(key), string(content), hash.ToString())
	}

	// Output:
	// Created author x2xi6sosy2yafpj3xx6zzocvfxzzdrwlkhaqew5v4pcccg2rtesa
	// Created document bq3ioxh7licvicknhuxl2in24qagzlewoaotgvm3xbr6nxwrv7ea
	// Inserted bafkr4ihasgdyqs6onufsjrmk5h5vcg2ud75u2iaokavwiulyg7wfno6fte
	// Keys:
	// "go" : "says hello" (hash: bafkr4ihasgdyqs6onufsjrmk5h5vcg2ud75u2iaokavwiulyg7wfno6fte)
	// Removing entry for author x2xi6sosy2yafpj3xx6zzocvfxzzdrwlkhaqew5v4pcccg2rtesa and prefix go.
	// Removed 1 entry
	// Keys:
}
